archival: same term sync-ing

VladLazar opened the pull request:
Previously, the archival STM used `persisted_stm::sync` in order to
ensure that its log is up to date before issuing new commands. This
works fine for leadership changes since they imply a term change.

However, the archiver may need to sync during the same term. We've seen
scenarios like this:

> Upload loop picks 4 candidates, uploads to cloud storage and
replicates a batch with 4 new segments. After the replication command
succeeds, the topic retention is updated. This re-starts the archiver.
When the archiver restarts it picks the same 4 segments and replicates
the addition commands again. Finally, the partition manifest detects
this, refuses the commands and prints the error log.

The previous sync method does not support such cases sync they happen in
the same term. To fix this, the archival STM now implements its own sync
method. If it detects that the term has changed, it will maintain the
pre-existing behaviour: sync up to the latest term. Otherwise, it will
use the cached last offset of the last replicated batch and wait on its
application.

Fixes https://github.com/redpanda-data/redpanda/issues/13753
## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes
### Bug Fixes
* Fix a bug where archival would not sync it's state correctly before proceeding with new segment uploads.
This could happen when updating the a topic's config.

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38194#018af6b6-a17b-468c-911e-290f99c403e8

VladLazar commented:
/ci-repeat 5
dt-log-level=debug
dt-repeat=10
tests/rptest/tests/tests/rptest/tests/tiered_storage_model_test.py

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -840,8 +979,26 @@ ss::future<> archival_metadata_stm::apply(const model::record_batch& b) {
+}
 
-    _manifest->advance_insync_offset(b.last_offset());
+void archival_metadata_stm::apply_prefix_truncate_batch(
```
oh, so it was handled inline inside the `apply` method? oh boy

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -664,6 +800,8 @@ ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
         co_return errc::replication_error;
     }
 
+    _last_replicated_offset = batch_last_offset;
```
I'm not quite following why do we need this. At this point the `batch_last_offset` is already applied to the STM. The `wait_no_throw` is waiting until the offset in first argument is applied to the STM. So the `_last_replicated_offset` should be equal to `_insync_offset`.

VladLazar replied:
You're correct. This is misplaced. It should be updated after the `replicate` call. Will move.

VladLazar replied:
fixed

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+     * ===================
+     *
+     * This function needs to handle two cases:
+     * Case 1: we have not synced since the last term (e.g. on leadership
```
Should this always be true because we're calling `sync` after the replica became a leader? And `sync` guarantees that all batches from all previous terms are applied IIRC.
cc @rystsov

VladLazar replied:
Right, that's what `persisted_stm::sync` does. The issue we are trying to prevent is related to stopping and starting the archiver in the same term.

Right now, we first signal the abort source and then wait for the gate to close. The following sequence of operations is possible:
1. archiver replicates some commands
2. replicate succeeds
3. we start waiting on apply
4. abort is requested by `ntp_archiver::stop`
5. waiting is aborted
6. we ended up in a case 2

An alternative solution is to make sure that `ntp_archiver::stop` leaves the STM in sync.
I think that's kind of brittle since someone can add a new path for replicating commands in the future
and not be aware of this requirement. It also relies very heavily on the correctness of replication semantics.
Hence why I picked the more generic approach of caching the latest replicated offset and waiting on it.

Lazin replied:
This looks like a solution to a problem that shouldn't exist. Why isn't it possible to avoid stopping the archiver?

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
 }
 
+ss::future<bool>
+archival_metadata_stm::sync(model::timeout_clock::duration timeout) {
```
Note for other reviewers: this method is called from the upload loop before any decisions are made. We want to make sure that the decisions are based on up to date state.

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+    const auto last_applied_offset = _manifest->get_insync_offset();
+
+    if (current_term == _insync_term && _last_replicated_offset) {
+        if (last_applied_offset < _last_replicated_offset) {
```
Given how the value is updated this can only be the case if we're calling this method concurrently. But this means that we have two active `ntp_archiver` instances. One is replicating something and about to step down and another one is starting uploads. We assume that there is always only one `ntp_archiver` and it's either reading data from the STM or replicating something.

VladLazar replied:
You're right. The place where `last_replicated_offset` is applied is not correct. Thanks for pointing out. Will fix.

VladLazar replied:
Fixed

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38281#018afbf3-bd91-4069-83e9-fb663c435876

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38281#018afbf3-bd97-4ca9-b0a9-809c9708a0f5

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38281#018afbf3-bda0-412e-863f-b6cfc98afbaf

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38281#018afc01-8147-4607-936e-3b93382c8bc2

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38281#018afc01-8146-41de-8631-7650f9752422

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38281#018afc01-813f-4e71-933b-1b9fbdf1dd6d

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38281#018afc01-813d-4510-96dc-06e6c39c6096

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38281#018afc01-813e-4063-82b9-8202877c9f36

VladLazar commented:
/ci-repeat 5
dt-log-level=debug
dt-repeat=10
tests/rptest/tests/tiered_storage_model_test.py

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38352#018aff57-1715-497f-b7f3-ea54c343b162

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38352#018aff57-1714-4748-ab4b-f398aa6191d5

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38352#018aff57-1715-454d-b9b1-7e1d0c098fac

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38352#018aff57-1714-4c1a-8ada-505d914305a7

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38352#018aff57-170c-418b-8894-502368085a9a

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38352#018aff57-170b-4c5c-a9c8-442eb715188f

VladLazar commented:
/ci-repeat 2
dt-log-level=debug
dt-repeat=20
tests/rptest/tests/tiered_storage_model_test.py

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38372#018b004d-9010-4bd4-8901-8fcdfa881ed0

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38372#018b004d-9006-42f4-9470-16a80898dd32

rystsov commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -650,6 +786,8 @@ ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
         co_return result.error();
     }
 
+    _last_replicated_offset = batch_last_offset;
```
I don't understand this, the offsets are assigned after the replication happens so in general case `batch.last_offset()` is undefined. Why don't you set `_last_replicated_offset` to `result.value().last_offset` after the replication happened?

VladLazar replied:
Good spot - thank you. Fixed.

andrwng commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+        }
+
+        if (
+          _last_replicated_offset
```
In this case, shouldn't we reset the replicated offset? Even though we've attempted to replicate in a previous term, there's no expectation for this STM to actually have applied it, because it's not guaranteed to have become Raft committed.

andrwng replied:
Actually I might be wrong about this -- probably worth double checking with someone from the replication team what the guarantee is here on success of `replicate()` if you haven't already

andrwng commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
 ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
   model::record_batch batch, ss::abort_source& as) {
-    auto current_term = _insync_term;
+    const auto batch_last_offset = batch.last_offset();
```
I thought the offset is assigned and serialized by Raft in the process of replicating. Perhaps we can get the resulting offset from the replicate() call?

VladLazar replied:
Yeah, that was a bug. Fixed now.

andrwng commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+                  current_term);
+                co_return false;
+            }
+        }
```
nit: if last_applied_offset >= _last_replicated_offset, should we co_return true immediately? Seems to reduce to the same thing, but without the below checks given there are no scheduling points with the above checks

andrwng commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+          _logger.debug,
+          "Syncing archival STM to latest term {}",
+          _raft->term());
+        const bool synced = co_await persisted_stm::sync(timeout);
```
nit: i'm wondering if we should call sync in both if/else blocks. Seems like its checks overlap with some in the above case (just the check for _last_replicated_offset is new, no?

VladLazar replied:
Yeah. I considered it too, but really there's no guarantee that we'll still be in the same term after the function returns. This means that the archiver needs to check anyways if it is still supposed to operate. So it'll call sync again in that case.

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38395#018b012c-16ff-4690-a636-fcafe9bf3e5c

VladLazar commented:
Failure was:
* https://github.com/redpanda-data/redpanda/issues/13456

VladLazar commented:
/ci-repeat 3
dt-log-level=debug
dt-repeat=10
tests/rptest/tests/tiered_storage_model_test.py

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38442#018b0448-c386-4850-a324-5216b8799b8d

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38442#018b0455-df83-491d-94b9-4d0c78f20b4b

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38442#018b0455-df83-4dde-b202-b36e6418cc8b

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38442#018b0455-df79-4e96-9ad5-71fc05502dde

VladLazar commented:
No "segment replaced with no change in size" failures occurred during the repeat.

andrwng reviewed (commented):
nits aside, lgtm. Would like someone from the replication team to help verify though

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -650,6 +785,8 @@ ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
         co_return result.error();
     }
 
+    _last_replicated_offset = result.value().last_offset;
```
I think it still doesn't solve the issue. Consider the following scenario:

* old archiver calls `replicate` and then `ssx::with_timeout_abortable`. The `_last_replicated_offset` contains old value;
* something stops the archiver midterm and `ssx::with_timeout_abortable` exits but the replication operation continues;
* new archiver calls `sync` and uses old `_last_replicated_offset` value;

andrwng replied:
Yea this is a good point. Replication that is attempted and appear to fail (eg because of a timeout) may still succeed, and I don't think we have a bulletproof way of knowing to wait for it.

I wondered briefly if we could cache the current committed offset before calling replicate, but it won't work given other calls to replicate may race with us.

VladLazar replied:
Had a chat with Evgeny about this. Our idea was to chain the update to `_last_replicated_offset` to the `replicate` future. That works, but thinking about it some more, I see another issue: the `replicate` fiber might not have finished by the time the archiver restarts so the new instance doesn't know it's supposed to wait.

In theory we could keep the `future` for the latest `replicate` in the STM and always wait on it in our `sync` implementation. Issue is that there may be multiple replication fibres in flight (from the segment merger and scrubber) and I'm not sure about the semantics of the "final" replicate. I've asked the replication team.

Lazin commented:
This is why I think we should concentrate on avoiding the race. Maybe also
think how we could validate the implementation. Maybe we need to learn
something from other teams.

On Fri, Oct 6, 2023 at 8:44 PM Andrew Wong ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/v/cluster/archival_metadata_stm.cc
> <https://github.com/redpanda-data/redpanda/pull/13896#discussion_r1349198802>
> :
>
> > @@ -650,6 +785,8 @@ ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
>          co_return result.error();
>      }
>
> +    _last_replicated_offset = result.value().last_offset;
>
> Yea this is a good point. Replication that is attempted and appear to fail
> (eg because of a timeout) may still succeed, and I don't think we have a
> bulletproof way of knowing to wait for it.
>
> I wondered briefly if we could cache the current committed offset before
> calling replicate, but it won't work given other calls to replicate may
> race with us.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/13896#discussion_r1349198802>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAWMNUFB6J2FEVSQPGAC6LX6BGRZAVCNFSM6AAAAAA5RH4ILGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTMNRSGYYTIOJSGM>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/149f18833530e91e646d682ac3e7a766b58eb8d0..636b16e73045df0f4919c64a4ae03dcdfbbb6b12):
* Reworked sync
* Added an archival stm fixture based on the new raft fixture
* Added delay injection to the new raft fixture
* Added a test that exercises same term syncing

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38717#018b1f3d-aa3b-49dd-838b-3165f1c54b52

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38717#018b1f3d-aa38-4b3a-8fc9-9ed3fa224e2f

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38717#018b1f61-8f9e-4dca-afd7-5e6b97bc3eba

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38717#018b1f61-8f9b-44b1-99a8-671e598795e2

andrwng commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+                std::ignore = co_await ss::with_timeout(
+                  before_replication + timeout, std::move(fut));
+            } catch (const ss::timed_out_error&) {
+                vlog(_logger.error, "Replication wait timed out");
+                co_return false;
```
I imagine it'd be unlikely, but if we get here, could we run into the same problem? Wouldn't the next iteration of the upload loop sync again, but this time not have a `_last_replicate` to wait on, and potentially proceed without waiting on the still-in-flight replication? On timeout, is there a way to set the future back?

VladLazar replied:
It's possible in theory. Evgeny pointed out the same thing. However, in practice I don't think this is a problem.
The timeout that we use is quite long 10s. If we don't replicate within that time we should lose leadership. At that point the term will have changed and we will fall through to `persisted_stm::sync` (old behaviour).

VladLazar replied:
> On timeout, is there a way to set the future back?

I think it might be possible to do it with a shared promise, but they're hard to get right.

andrwng reviewed (commented):
Nice tests!

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+    const auto current_term = _insync_term;
+
+    ss::promise<result<raft::replicate_result>> replication_promise;
+    _last_replicate = replication_promise.get_future();
```
Maybe add assertion that checks that `_last_replicate` is equal to `std::nullopt`? We're calling `sync` every time before calling `do_relicate_commands` so the invariant `_last_replciate == std::nullopt` always holds but it's better to make it explicit.

Lazin replied:
Another important thing is that previously, we could call `do_replicate_commands` independently from `sync`. The sync was called just in case we're at the start of the term I believe (don't quite remember to be honest). But now you have to cal `sync` before `do_replicate_commands` to clean `_last_replicate`. Otherwise it will barf after the promise is set. So maybe it makes sense to add some comment that explains this.

VladLazar replied:
Right. `sync` should be called before `do_replicate_commands` and both. calls should be under the stm lock.

VladLazar replied:
Added assertions to `do_replicate_commands`.

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+            const auto after_replication = model::timeout_clock::now();
+            timeout -= after_replication - before_replication;
+        } else {
+            co_await fut.discard_result();
```
I think here you don't need to add a scheduling point. You can just call `fut.get().discard_result()`. Call to `get()` will propagate exception if the replication has failed.

VladLazar replied:
Makes sense. I realised that we probably don't want to propagate the exception here. It's fine for replication to fail. `sync` should continue in that case.

VladLazar replied:
Fixed by using `ignore_ready_future` if the future was ready. Also updated the other branch to ignore any exceptions from `consensus::replicate` and proceed.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/78fb37933c04ffa98bf2b8ea69cd8d119d1653c8..f55b6bd74fb0bd2cd02fa245f09d28cde0fda7b0):
* Make invariants for `do_replicate_commands` explicit via assertions
* Handle exceptional returns from `replicate`
* Prevent over/underflow of timeout

VladLazar commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+            }
+
+            const auto after_replication = model::timeout_clock::now();
+            timeout -= after_replication - before_replication;
```
Fixed.

VladLazar commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+    // Note that we throw away the result and use the Raft's committed offset
+    // below. This deals with both success and failure paths.
+    if (_last_replicate) {
+        auto fut = std::move(_last_replicate.value());
```
Done.

Lazin commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,13 +620,144 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+                vlog(
+                  _logger.debug,
+                  "Replication failed for archival STM command. Continuing "
+                  "sync.");
```
Why is it OK to continue? If we received some error from raft does this mean that we lost leadership and need to just exit?

VladLazar replied:
I don't think it really matters what we do on the replication failed case. If `replicate` returns an exceptional future, that should propagate and trigger the upload loop to check terms again. If `replicate` fails with an error code, we step and the rest of this function is a no-op.

Still, maybe it's easier to reason about if we return a sync failure in these cases. I'll do that.

VladLazar replied:
Changed it to exit early and added another test which simulates a replication failure.

VladLazar commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+
+        if (!synced) {
+            vlog(
+              _logger.warn,
```
I think warn is fine, but we should create an alert such that we know if it happens too frequently (say 10x per minute).

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38811#018b2480-d7be-4e7e-90e8-3760587243c2

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2a10fc80fab938b39e49ceef2456998325259725..8b66d00562b6226f38d8761942b30829b151caa0):
* fix oss build

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38917#018b299e-3d2a-4318-9677-948004352e19

VladLazar commented:
/ci-repeat 5

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/8b66d00562b6226f38d8761942b30829b151caa0..efb20b7ab30cfe6b95f5cac9893d23ead7707d6f):
* Fixed a race condition in the raft fixture shutdown which surfaced during the ci-repeat
* Demote a log line from error to warn.

VladLazar commented:
/ci-repeat 3

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/39303#018b480b-b60b-45dd-9433-3218b800f976

VladLazar commented:
All the failures seem known. Maybe we should merge. We can always revert if it turns out to cause breakage. @Lazin could you please re-approve?

mmaslankaprv commented on src/v/raft/tests/raft_fixture.cc:
```diff
@@ -281,8 +292,24 @@ in_memory_test_protocol::dispatch(model::node_id id, ReqT req) {
+
+        if (auto iter = _failures.find(msg_type); iter != _failures.end()) {
+            auto& fail = iter->second;
+            co_await ss::visit(fail, [this, msg_type](response_delay& f) {
```
nit: i would probably use a command pattern here i.e. instead of a visit encapsulate the failure injection logic inside the operation itself, but it is a matter of taste

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/39459#018b4e82-2984-4602-a640-dfe5fae9eb3f: "rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move_x_core.replication_factor=3.unclean_abort=False.recovery=restart_recovery.compacted=False"

VladLazar commented:
I don't think the failure above is related. The archival STM is not started in the failed test.

VladLazar commented:
Failures are:
* https://github.com/redpanda-data/redpanda/issues/14351
* https://github.com/redpanda-data/redpanda/issues/14254

vbotbuildovich commented:
/backport v23.2.x

vbotbuildovich commented:
Failed to create a backport PR to v23.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-13896-v23.2.x-800 remotes/upstream/v23.2.x
git cherry-pick -x 181ac6c06bf909c3815028913849e91d7287dedd e87c4ecb15b3df944b3b31881e0ad963d7067ee7 68561c071b58d241198915a96ece270e201f7543 a6789de2b261b1e983714843e388107ad1a279b2 d65f0ac98847ed3ca980f326f32bda8a019f1eb4 a277beaf232c12883ad514fd9c23e402787a83e3 de6e7f4ede37957444a68a57a488962717f66f99 04730258384db2e37afb53338f3d5703a03d3b04
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/6613485148).
